### ISSUE TITLE
Displaying cheat sheets with markdown

### ISSUE THREAD

**Issue by seatrout** [cheat/cheat#488]
This may be impossible, but I would like to be able to write cheat sheets in Markdown and have them display properly in a terminal afterwards. This would definitely make long ones easier to read. 

When I try adding a YAML frontmatter header, as the docs suggest, 
```
---
syntax: markdown
---
```

Nothing happens, except that the directive I have added displays in plain text. 

Is this something the chroma formatter is meant to be able to handle, or am I asking for something impossible?

**Comment by chrisallenlane** [comment]
Hmm. Would you mind pasting or linking your cheatsheet here? This should work already. I have a few personal markdown-formatted sheets that I use, and they seem to work fine.

**Comment by chrisallenlane** [comment]
One thing to check: make sure that you are, in fact, using `cheat` version 3 rather than 2. (`cheat --version` will tell you which you're running.)

While developing version 3, a few times I ended up executing version 2 a few times without realizing it, causing some confusion.

**Comment by seatrout** [comment]
Ah. You're right. I was still using 2.4

The problem is that I can't upgrade it -- at least, when I download the
3.01 executable -- cheat-darwin-amd64 --  my mac refuses to run it (under
Catalina). Says it "cannot be opened because the developer cannot be
verified" and offers either to move it to the bit or to cancel.

I suspect this is a problem I could have avoided by not upgrading, but will
consult Google


-------------------
Andrew Brown
Leader writer,
The Guardian
mobile: +44-(0)7968-945618

**Comment by chrisallenlane** [comment]
Yeah, the new version of `cheat` is backwards-incompatible with prior versions. It's unfortunate, but was necessary.

Run this:
`cheat --help`

You'll see this line under "Examples":

```
To initialize a config file:
  mkdir -p ~/.config/cheat && cheat --init > ~/.config/cheat/conf.yml
```

Run that to create a config file. You'll then want to edit that file to your preferences. It's pretty well-commented and self-explanatory.

One note regarding upgrading `cheat` going forward: the `brew` maintainers are doing an excellent job of keeping their package up-to-date, so you should actually be able to install `cheat` like this:

```
brew install cheat
```

(That's assuming that you use `brew`, of course.)

That might be more convenient that repeatedly downloading the binary

**Comment by chrisallenlane** [comment]
Yeah, the new version of `cheat` is backwards-incompatible with prior versions. It's unfortunate, but was necessary.

Run this:
`cheat --help`

You'll see this line under "Examples":

```
To initialize a config file:
  mkdir -p ~/.config/cheat && cheat --init > ~/.config/cheat/conf.yml
```

Run that to create a config file. You'll then want to edit that file to your preferences. It's pretty well-commented and self-explanatory.

One note regarding upgrading `cheat` going forward: the `brew` maintainers are doing an excellent job of keeping their package up-to-date, so you should actually be able to install `cheat` like this:

```
brew install cheat
```

(That's assuming that you use `brew`, of course.)

That might be more convenient that repeatedly downloading the binary, because development is going to be very active for the next few days/weeks.

**Comment by chrisallenlane** [comment]
The "front-matter" goes in individual cheatsheets. You don't want to change the structure of the config file at all - just fill in its values.

>  I enclose an example of a file that

This didn't seem to make it through. Feel free to post it here directly if you'd like if the advice above doesn't solve your problem.

**Comment by seatrout** [comment]
Oh, this is frustrating. And thank you for being so patient.

Here is a paste of the troublesome file

---
syntax: markdown
tags: [ personal ]
---


## Loops
`
for name ... [ in word ... ] term do list done
`

>   Expand the list of words, and set the parameter name to each of them in
turn, executing list each time. If the ‘in word’ is omitted, use the
positional parameters instead of the words.

> The term consists of one or more newline or ; which terminate the words,
and are optional when the ‘in word’ is omitted.

> More than one parameter name can appear before the list of words. If N
names are given, then on each execution of the loop the next N words are
assigned to the corresponding parameters. If there are more names than
remaining words, the remaining parameters are each set to the empty string.
Execution of the loop ends when there is no remaining word to assign to the
first name. It is only possible for in to appear as the first name in the
list, else it will be treated as marking the end of the list.

`
for (( [expr1] ; [expr2] ; [expr3] )) do list done
`

>The arithmetic expression expr1 is evaluated first (see Arithmetic
Evaluation). The arithmetic expression expr2 is repeatedly evaluated until
it evaluates to zero and when non-zero, list is executed and the arithmetic
expression expr3 evaluated. If any expression is omitted, then it behaves
as if it evaluated to 1.

`
while list do list done
`

>Execute the do list as long as the while list returns a zero exit status.

`
until list do list done
`

>Execute the do list as long as until list returns a nonzero exit status.

# command line history expansion

`
0
`
The first input word (command).

`
n
`
The nth argument.

`^`
The first argument. That is, 1.

`$`
The last argument.


# filename globbing
>After the optional word designator, you can add a sequence of one or more
of the following modifiers, each preceded by a ‘:’. These modifiers also
work on the result of filename generation and parameter expansion, except
where noted.

`e`
>Remove all but the part of the filename extension following the ‘.’; see
the definition of the filename extension in the description of the r
modifier below. Note that according to that definition the result will be
empty if the string ends with a ‘.’.

`h`

>Remove a trailing pathname component, leaving the head. This works like
‘dirname’.

`l`   Convert the words to all lowercase.

`r`

>Remove a filename extension leaving the root name. Strings with no
filename extension are not altered. A filename extension is a ‘.’ followed
by any number of characters (including zero) that are neither ‘.’ nor ‘/’
and that continue to the end of the string. For example, the extension of
‘foo.orig.c’ is ‘.c’, and ‘dir.c/foo’ has no extension.


`t`

Remove all leading pathname components, leaving the tail. This works like
‘basename’.

`u`

Convert the words to all uppercase.

### Glob Qualifiers ###

>Patterns used for filename generation may end in a list of qualifiers
enclosed in parentheses. The qualifiers specify which filenames that
otherwise match the given pattern will be inserted in the argument list.

`/`
directories

`F`

> ‘full’ (i.e. non-empty) directories. Note that the opposite sense (^F)
expands to empty directories and all non-directories. Use (/^F) for empty
directories.

`.`
plain files

`@`  symbolic links

`a[Mwhms][-|+]n`

>files accessed exactly n days ago. Files accessed within the last n days
are selected using a negative value for n `(-n)`. Files accessed more than
n days ago are selected by a positive n value `(+n)`. Optional unit
specifiers `‘M’, ‘w’, ‘h’, ‘m’` or `‘s’` (e.g. `‘ah5’`) cause the check to
be performed with months (of 30 days), weeks, hours, minutes or seconds
instead of days, respectively. An explicit `‘d’` for days is also allowed.

`L[+|-]n`

>files less than n bytes `(-)`, more than n bytes `(+)`, or exactly n bytes
in length. If this flag is directly followed by a size specifier `‘k’
(‘K’), ‘m’ (‘M’), or ‘p’ (‘P’)` (e.g. `‘Lk-50’`) the check is performed
with kilobytes, megabytes, or blocks (of 512 bytes) instead. (On some
systems additional specifiers are available for gigabytes, ‘g’ or ‘G’, and
terabytes, ‘t’ or ‘T’.)

`oc`

>specifies how the names of the files should be sorted. If c is n they are
sorted by name; if it is L they are sorted depending on the size (length)
of the files; if l they are sorted by the number of links; if a, m, or c
they are sorted by the time of the last access, modification, or inode
change respectively; if d, files in subdirectories appear before those in
the current directory at each level of the search — this is best combined
with other criteria, for example ‘odon’ to sort on names for files within
the same directory

`Oc`

> like ‘o’, but sorts in descending order; i.e. ‘*(^oc)’ is the same as
‘*(Oc)’ and ‘*(^Oc)’ is the same as ‘*(oc)’; ‘Od’ puts files in the current
directory before those in subdirectories at each level of the search.


-------------------
Andrew Brown
Leader writer,
The Guardian
mobile: +44-(0)7968-945618

**Comment by chrisallenlane** [comment]
@seatrout 

Hmm... Your cheatsheet actually displays just fine on my machine. I see nice colorized headings and such.

One more time: what's your output from this command?

```sh
cheat --version
```

I'm wondering if you still have a version 2 on your path somewhere.

Feel free to paste your cheat config file here as well.

**Comment by seatrout** [comment]
cheat -- version gives me 3.0.1

I get the colourised headings and code. though the colours of the headings
are not very helpful (see screenshot).
But I don't get the text indented as it should be, and the markdown
characters for headings and so on remain visible in the terminal.

See the enclosed screenshot, with a terminal window in front of the VScode
preview

I think the problem lies in the underlying library, and the randomness with
which it assigns elements to colours. I have fooled around with a lot of
different "style" settings in conf.yml and none of them do any more
formatting than simply colouring elements differently; also, none seem to
have a setting which makes all the different elements readable. I know all
this is better than what we had, and I don't want to complain groundlessly.
Maybe it's just another example of the capacity of software to frustrate us
by promising a little more than it can deliver in this case full-on
markdown rendering.

A


[image: Screenshot 2019-10-29 at 20.38.15.png]



-------------------
Andrew Brown
Leader writer,
The Guardian
mobile: +44-(0)7968-945618

**Comment by seatrout** [comment]
For what it's worth, here is also my current conf.yml file

---
# The editor to use with 'cheat -e <sheet>'. Defaults to $EDITOR or $VISUAL.
editor: vim

# Should 'cheat' always colorize output?
colorize: true

# Which 'chroma' colorscheme should be applied to the output?
# Options are available here:
#   https://github.com/alecthomas/chroma/tree/master/styles
style: solarized-dark

# Which 'chroma' "formatter" should be applied?
# One of: "terminal", "terminal256", "terminal16m"
formatter: terminal16m

# The paths at which cheatsheets are available. Tags associated with a cheatpath
# are automatically attached to all cheatsheets residing on that path.
#
# Whenever cheatsheets share the same title (like 'tar'), the most local
# cheatsheets (those which come later in this file) take precedent over the
# less local sheets. This allows you to create your own "overides" for
# "upstream" cheatsheets.
#
# But what if you want to view the "upstream" cheatsheets instead of your own?
# Cheatsheets may be filtered via 'cheat -f <tag>' in combination with other
# commands. So, if you want to view the 'tar' cheatsheet that is tagged as
# 'community' rather than your own, you can use: cheat tar -f community
cheatpaths:

  # Paths that come earlier are considered to be the most "global", and will
  # thus be overridden by more local cheatsheets. That being the case, you
  # should probably list community cheatsheets first.
  #

# Note that the paths and tags listed below are just examples. You may freely
  # change them to suit your needs.
  - name: community
    path: ~/.cheat/community
    tags: [ community ]
    readonly: true


# Maybe your company or department maintains a repository of cheatsheets as
  # well. It's probably sensible to list those second.
  # - name: work
  #   path: ~/.cheat/work
  #   tags: [ work ]
  #   readonly: false

  # If you have personalized cheatsheets, list them last. They will take
  # precedence over the more global cheatsheets.
  - name: personal
    path: ~/.cheat/personal
    tags: [ personal ]
    readonly: false



-------------------
Andrew Brown
Leader writer,
The Guardian
mobile: +44-(0)7968-945618

**Comment by zeitchef** [comment]
@chrisallenlane I'm on version 3.0.3 and am running into the exact same issue - no Markdown
rendering.

**Comment by chrisallenlane** [comment]
I confirmed with @seatrout via another channel that `cheat` is behaving as expected. I'll elaborate on the situation here for other users that may encounter this issue in the future.

`cheat` (currently) will apply syntax-highlighting to markdown, but it does not _render_ markdown via any kind of layout engine. (I'm not aware of whether any such engine is conveniently available for use.) As a consequence, `cheat` will colorize markdown documents, but it will not format them as one would expect to view them on Github.

While possibly unfortunate, this is the current intended behavior. That being the case, I'm going to close this issue as resolved.

With that said, the ability to render markdown could be useful. I have opened #494 to track this proposed feature going forward.
